app/scripts/lib/WeakRefObjectMap.ts

```diff
@@ -0,0 +1,74 @@
+type WeakRefCompatibleObject<T> = {
+  [P in keyof T]: T[P] extends object ? WeakRef<T[P]> : never;
+};
+
+export class WeakRefObjectMap {
+  map: Map<string, WeakRefCompatibleObject<Record<string, object>>>;
+
+  constructor() {
+    this.map = new Map();
+  }
+
+  set(key: string, value: Record<string, object>) {
+    const weakReffedValueObj = {} as WeakRefCompatibleObject<
+      Record<string, object>
+    >;
+    Object.keys(value).forEach((valueKey) => {
+      weakReffedValueObj[valueKey] = new WeakRef(value[valueKey]);
+    });
+    this.map.set(key, weakReffedValueObj);
+  }
+
+  get(key: string) {
```

Comment:
Nit: I'd recommend providing explicit return types for any of the methods that return a value. This aids readability and ensures the return type doesn't change accidentally (this can be easy to do accidentally if we're relying on type inference)

See here for more on that recommendation: https://github.com/MetaMask/contributor-docs/blob/85e06a3097ffdacb7b2fbda9d074955582660181/docs/typescript.md#for-functions-and-methods-provide-explicit-return-types

Reply:
done [here](https://github.com/MetaMask/metamask-extension/pull/23525/commits/a3770013832099fc91e51e849f3ff399b556d4f8) and [here](https://github.com/MetaMask/metamask-extension/pull/23525/commits/94c4976528bb0f83c9815acd22be4a45fa4f5646)